agent/agent-profiler/agent-alerting/src/main/java/com/microsoft/applicationinsights/alerting/analysis/aggregations/windowed/WindowedAggregation.java

```diff
@@ -0,0 +1,112 @@
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.microsoft.applicationinsights.alerting.analysis.aggregations.windowed;
+
+import com.microsoft.applicationinsights.alerting.analysis.TimeSource;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/** Applies a time window to data held in 1 second buckets. I.e holds the last n seconds of data. */
+public class WindowedAggregation<T extends BucketData<U>, U> {
+  public static final int BUCKET_DURATION_SECONDS = 2;
+  private final long windowLengthInSec;
+  private final TimeSource timeSource;
+
+  private final Object bucketLock = new Object();
+  private final List<WindowedAggregationBucket<T, U>> buckets =
+      Collections.synchronizedList(new ArrayList<>());
+  private WindowedAggregationBucket<T, U> currentBucket;
+  private final Supplier<T> bucketFactory;
+
+  // Determines if the current bucket that is in the process of being calculated is included
+  // in the returned data
+  private final boolean trackCurrentBucket;
+
+  public WindowedAggregation(
+      long windowLengthInSec,
+      TimeSource timeSource,
+      Supplier<T> bucketFactory,
+      boolean trackCurrentBucket) {
+    this.windowLengthInSec = windowLengthInSec;
+    this.timeSource = timeSource;
+    this.bucketFactory = bucketFactory;
+    this.trackCurrentBucket = trackCurrentBucket;
+  }
+
+  public void update(U breached) {
+    getBucket().update(breached);
+  }
+
+  public List<T> getData() {
+    Instant now = timeSource.getNow();
+    Instant cutoff = now.minusSeconds(windowLengthInSec);
+    gcBuckets(cutoff);
+    return buckets.stream().map(WindowedAggregationBucket::getData).collect(Collectors.toList());
+  }
+
+  private WindowedAggregationBucket<T, U> getBucket() {
+    synchronized (bucketLock) {
+      Instant now = timeSource.getNow();
+
+      if (currentBucket == null) {
+        currentBucket =
+            new WindowedAggregationBucket<>(
+                now.plusSeconds(BUCKET_DURATION_SECONDS), bucketFactory.get());
+        if (trackCurrentBucket) {
+          buckets.add(currentBucket);
+        }
+      }
+
+      if (currentBucket.bucketEnd.isBefore(now)) {
+        // Gone past end of current bucket, add it to the array
+        Instant cutoff = now.minusSeconds(windowLengthInSec);
+        gcBuckets(cutoff);
+
+        if (!trackCurrentBucket) {
+          buckets.add(currentBucket);
+        }
+
+        currentBucket =
+            new WindowedAggregationBucket<>(
+                now.plusSeconds(BUCKET_DURATION_SECONDS), bucketFactory.get());
+
+        if (trackCurrentBucket) {
+          buckets.add(currentBucket);
+        }
+      }
+
+      return currentBucket;
+    }
+  }
+
+  private void gcBuckets(Instant cutoff) {
```

Comment:
what does `gc` mean here?

Reply:
was using to to mean garbage collect, renamed it to `removeBucketsBeforeCutoff`